lab3van/util-linux/misc-utils/rename.1
.\" Written by Andries E. Brouwer (aeb@cwi.nl)
.\" Placed in the public domain
.\"
.TH RENAME 1 "June 2011" "util-linux" "User Commands"
.SH NAME
rename \- rename files
.SH SYNOPSIS
.B rename
<<<<<<< HEAD
.RI [ options ] " expression replacement file" ...
=======
[options]
.IR "expression replacement file" ...
>>>>>>> master-vanilla
.SH DESCRIPTION
.B rename
will rename the specified files by replacing the first occurrence of
.I expression
in their name by
.IR replacement .
.SH OPTIONS
.TP
<<<<<<< HEAD
\fB\-v\fR, \fB\-\-verbose\fR
Give visual feedback which files where renamed, if any.
.TP
\fB\-V\fR, \fB\-\-version\fR
Display version information and exit.
.TP
\fB\-h\fR, \fB\-\-help\fR
=======
.BR \-s , " \-\-symlink"
Do not rename a symlink but its target.
.TP
.BR \-v , " \-\-verbose"
Show which files where renamed, if any.
.TP
.BR \-V , " \-\-version"
Display version information and exit.
.TP
.BR \-h , " \-\-help"
>>>>>>> master-vanilla
Display help text and exit.
.SH EXAMPLES
Given the files
.IR foo1 ", ..., " foo9 ", " foo10 ", ..., " foo278 ,
the commands
.RS
.PP
.nf
rename foo foo0 foo?
rename foo foo0 foo??
.fi
.PP
.RE
will turn them into
.IR foo001 ", ..., " foo009 ", " foo010 ", ..., " foo278 .
And
.RS
.PP
.nf
rename .htm .html *.htm
.fi
.PP
.RE
will fix the extension of your html files.
.SH WARNING
The renaming has no safeguards.  If the user has permission to rewrite file names,
the command will perform the action without any questions.  For example, the
result can be quite drastic when the command is run as root in the /lib
directory.  Always make a backup before running the command, unless you truly
know what you are doing.
<<<<<<< HEAD
.SH "SEE ALSO"
.BR mmv (1),
=======
.SH "EXIT STATUS"
.RS
.PD 0
.TP
.B 0
all requested rename operations were successful
.TP
.B 1
all rename operations failed
.TP
.B 2
some rename operations failed
.TP
.B 4
nothing was renamed
.TP
.B 64
unanticipated error occurred
.PD
.RE
.SH "SEE ALSO"
>>>>>>> master-vanilla
.BR mv (1)
.SH AVAILABILITY
The rename command is part of the util-linux package and is available from
ftp://ftp.kernel.org/pub/linux/utils/util-linux/.


debian/changelog
util-linux (2.27.160-1) unstable; urgency=low

  * Initial release (Closes: #nnnn)  <nnnn is the bug number of your ITP>

 -- roma <newangel88@mail.ru>  Thu, 24 Dec 2015 13:07:17 +0400


lab3van/util-linux/misc-utils/rename.c
/*
 * rename.c - aeb 2000-01-01
 *
--------------------------------------------------------------
#!/bin/sh
if [ $# -le 2 ]; then echo call: rename from to files; exit; fi
FROM="$1"
TO="$2"
shift
shift
for i in $@; do N=`echo "$i" | sed "s/$FROM/$TO/g"`; mv "$i" "$N"; done
--------------------------------------------------------------
 * This shell script will do renames of files, but may fail
 * in cases involving special characters. Here a C version.
 */
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <errno.h>
#include <getopt.h>
<<<<<<< HEAD
=======
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
>>>>>>> master-vanilla

#include "nls.h"
#include "xalloc.h"
#include "c.h"
<<<<<<< HEAD

static int do_rename(char *from, char *to, char *s, int verbose)
{
	char *newname, *where, *p, *q;
	int flen, tlen, slen;

	where = strstr(s, from);
	if (where == NULL)
		return 0;

	flen = strlen(from);
	tlen = strlen(to);
	slen = strlen(s);
	newname = xmalloc(tlen + slen + 1);

	p = s;
	q = newname;
=======
#include "closestream.h"

#define RENAME_EXIT_SOMEOK	2
#define RENAME_EXIT_NOTHING	4
#define RENAME_EXIT_UNEXPLAINED	64

static int string_replace(char *from, char *to, char *s, char *orig, char **newname)
{
	char *p, *q, *where;

	where = strstr(s, from);
	if (where == NULL)
		return 1;
	p = orig;
	*newname = xmalloc(strlen(orig) + strlen(to) + 1);
	q = *newname;
>>>>>>> master-vanilla
	while (p < where)
		*q++ = *p++;
	p = to;
	while (*p)
		*q++ = *p++;
<<<<<<< HEAD
	p = where + flen;
	while (*p)
		*q++ = *p++;
	*q = 0;

	if (rename(s, newname) != 0)
		err(EXIT_FAILURE, _("renaming %s to %s failed"),
		    s, newname);
	if (verbose)
		printf("`%s' -> `%s'\n", s, newname);

	free(newname);
	return 1;
=======
	p = where + strlen(from);
	while (*p)
		*q++ = *p++;
	*q = 0;
	return 0;
}

static int do_symlink(char *from, char *to, char *s, int verbose)
{
	char *newname = NULL, *target = NULL;
	int ret = 1;
	struct stat sb;

	if (lstat(s, &sb) == -1) {
		warn(_("stat of %s failed"), s);
		return 2;
	}
	if (!S_ISLNK(sb.st_mode)) {
		warnx(_("%s: not a symbolic link"), s);
		return 2;
	}
	target = xmalloc(sb.st_size + 1);
	if (readlink(s, target, sb.st_size + 1) < 0) {
		warn(_("%s: readlink failed"), s);
		free(target);
		return 2;
	}
	target[sb.st_size] = '\0';
	if (string_replace(from, to, target, target, &newname))
		ret = 0;
	else if (0 > unlink(s)) {
		warn(_("%s: unlink failed"), s);
		ret = 2;
	} else if (symlink(newname, s) != 0) {
		warn(_("%s: symlinking to %s failed"), s, newname);
		ret = 2;
	}
	if (verbose && ret == 1)
		printf("%s: `%s' -> `%s'\n", s, target, newname);
	free(newname);
	free(target);
	return ret;
}

static int do_file(char *from, char *to, char *s, int verbose)
{
	char *newname = NULL, *file=NULL;
	int ret = 1;

	if (strchr(from, '/') == NULL && strchr(to, '/') == NULL)
		file = strrchr(s, '/');
	if (file == NULL)
		file = s;
	if (string_replace(from, to, file, s, &newname))
		return 0;
	else if (rename(s, newname) != 0) {
		warn(_("%s: rename to %s failed"), s, newname);
		ret = 2;
	}
	if (verbose && ret == 1)
		printf("`%s' -> `%s'\n", s, newname);
	free(newname);
	return ret;
>>>>>>> master-vanilla
}

static void __attribute__ ((__noreturn__)) usage(FILE * out)
{
<<<<<<< HEAD
	fputs(_("\nUsage:\n"), out);
	fprintf(out,
	      _(" %s [options] expression replacement file...\n"),
		program_invocation_short_name);

	fputs(_("\nOptions:\n"), out);
	fputs(_(" -v, --verbose    explain what is being done\n"
		" -V, --version    output version information and exit\n"
		" -h, --help       display this help and exit\n\n"), out);

=======
	fputs(USAGE_HEADER, out);
	fprintf(out,
	      _(" %s [options] <expression> <replacement> <file>...\n"),
		program_invocation_short_name);

	fputs(USAGE_SEPARATOR, out);
	fputs(_("Rename files.\n"), out);

	fputs(USAGE_OPTIONS, out);
	fputs(_(" -v, --verbose    explain what is being done\n"), out);
	fputs(_(" -s, --symlink    act on the target of symlinks\n"), out);
	fputs(USAGE_SEPARATOR, out);
	fputs(USAGE_HELP, out);
	fputs(USAGE_VERSION, out);
	fprintf(out, USAGE_MAN_TAIL("rename(1)"));
>>>>>>> master-vanilla
	exit(out == stderr ? EXIT_FAILURE : EXIT_SUCCESS);
}

int main(int argc, char **argv)
{
	char *from, *to;
<<<<<<< HEAD
	int i, c, verbose = 0;
=======
	int i, c, ret = 0, verbose = 0;
	int (*do_rename)(char *from, char *to, char *s, int verbose) = do_file;
>>>>>>> master-vanilla

	static const struct option longopts[] = {
		{"verbose", no_argument, NULL, 'v'},
		{"version", no_argument, NULL, 'V'},
		{"help", no_argument, NULL, 'h'},
<<<<<<< HEAD
=======
		{"symlink", no_argument, NULL, 's'},
>>>>>>> master-vanilla
		{NULL, 0, NULL, 0}
	};

	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
<<<<<<< HEAD

	while ((c = getopt_long(argc, argv, "vVh", longopts, NULL)) != -1)
=======
	atexit(close_stdout);

	while ((c = getopt_long(argc, argv, "vsVh", longopts, NULL)) != -1)
>>>>>>> master-vanilla
		switch (c) {
		case 'v':
			verbose = 1;
			break;
<<<<<<< HEAD
		case 'V':
			printf(_("%s from %s\n"),
			       program_invocation_short_name,
			       PACKAGE_STRING);
=======
		case 's':
			do_rename = do_symlink;
			break;
		case 'V':
			printf(UTIL_LINUX_VERSION);
>>>>>>> master-vanilla
			return EXIT_SUCCESS;
		case 'h':
			usage(stdout);
		default:
			usage(stderr);
		}

	argc -= optind;
	argv += optind;

	if (argc < 3) {
<<<<<<< HEAD
		warnx("not enough arguments");
=======
		warnx(_("not enough arguments"));
>>>>>>> master-vanilla
		usage(stderr);
	}

	from = argv[0];
	to = argv[1];

	for (i = 2; i < argc; i++)
<<<<<<< HEAD
		do_rename(from, to, argv[i], verbose);

	return EXIT_SUCCESS;
=======
		ret |= do_rename(from, to, argv[i], verbose);

	switch (ret) {
	case 0:
		return RENAME_EXIT_NOTHING;
	case 1:
		return EXIT_SUCCESS;
	case 2:
		return EXIT_FAILURE;
	case 3:
		return RENAME_EXIT_SOMEOK;
	default:
		return RENAME_EXIT_UNEXPLAINED;
	}
>>>>>>> master-vanilla
}
